pkg/util/util.go

```diff
@@ -232,29 +235,44 @@ func ParseMachinesYaml(file string) ([]*clusterv1.Machine, error) {
 	defer reader.Close()
 
 	decoder := yaml.NewYAMLOrJSONDecoder(reader, 32)
-	machineList, err := decodeMachineLists(decoder)
 
+	var (
+		bytes    [][]byte
+		machines = []clusterv1.Machine{}
+	)
+
+	// TODO: use the universal decoder instead of doing this.
+	bytes, err = decodeClusterV1Kinds(decoder, "MachineList")
 	if err != nil {
+		if isMissingKind(err) {
+			err = fmt.Errorf(MachineListFormatDeprecationError)
+		}
 		return nil, err
 	}
+	// TODO: this is O(n^2) and must be optimized
+	for _, ml := range bytes {
+		var machineList clusterv1.MachineList
+		err = json.Unmarshal(ml, &machineList)
+		if err != nil {
```

Comment:
Prefer using `if err := json.Unmarshal(ml, &machineList); err !=` which scopes `err` to the following block.